### ISSUE TITLE
Custom runtime grandpa_authorities call leads to panic and exit

### ISSUE THREAD

**Issue by MatthewDarnell** [paritytech/substrate#10401]
I'm working on a custom barebones C wasm runtime and am trying to get a valid function for `GrandpaApi_grandpa_authorities`. It has the wasm signature `(i32, i32) -> i64`

```
signed long long int GrandpaApi_grandpa_authorities(signed int one, signed int two) {
  unsigned long long int f = 25;
  f <<= 32;
  f |= 0x0a;
  return f;
}
```

I'm not sure whether or not this is a valid authority list, (I added an info log that shows the data being passed: `2021-12-01 10:27:10 AuthorityList::genesis() 0x16da2ee30`

When it runs, it crashes:

```
Thread 'main' panicked at 'genesis authorities is non-empty; all weights are non-zero; qed.', /Users/matthew/Documents/substrate/client/finality-grandpa/src/aux_schema.rs:392

This is a bug. Please report it at:

	support.anonymous.an
```

**Comment by bkchr** [comment]
As the error says, you can not return an empty vector there. What you are decoding there, is clearly not a correct vector with the required values.

**Comment by MatthewDarnell** [comment]
> As the error says, you can not return an empty vector there. What you are decoding there, is clearly not a correct vector with the required values.

I realize it's not a correct vector. I'm assuming that the correct byte layout of looks like a `Vec<( [u8; 32], u64)>`. I just thought that it's a bug because it self reports as a bug and crashes the node.

**Comment by bkchr** [comment]
> I just thought that it's a bug because it self reports as a bug and crashes the node

That is because it is a panic/a hard constraint.

**Comment by MatthewDarnell** [comment]
> > I just thought that it's a bug because it self reports as a bug and crashes the node
> 
> That is because it is a panic/a hard constraint.

So it's expected behavior? That's fine, I will close the issue if so.

Btw, I'm just trying to find any return value from the function that passes. The function signature in the wasm says it returns an `i64` but it seems too small to return a Vector containing this tuple.

**Comment by bkchr** [comment]
https://github.com/paritytech/substrate/blob/master/primitives/core/src/lib.rs#L400-L420

All runtime api functions return an `i64`. This `i64` is the pointer + length of the SCALE encoded return type. The function I listed above is the one that creates this return type. This should help you to create the correct value.

**Comment by bkchr** [comment]
> So it's expected behavior? That's fine, I will close the issue if so.

Yes that is expected. Aka we require that genesis always returns a non-empty vec.

**Comment by pepyakin** [comment]
In Parity we stick to this "proof or remove" policy about panics. That is, the panics should either come with a proof that they will never-ever fire or removed completely (either by type-system or returning a `Result` or w/e).

Correct me if I am wrong, but the proof there is a complete bogus. It, apparently, relies on the `GrandpaApi_grandpa_authorities` to return a non-empty list. The client code cannot rely on this particular behavior, because, well as was proven by the OP's counter-example, nothing stops runtime to return anything. Hence, the proof is wrong and the code needs to be changed.

↑ This said with my pedantic hat on.
↓ That said with my pragmatic hat on.

I understand that fixing this may lead to massive refactorings and won't give any benefit because all sane runtimes do return some values from here. If that's the case, I think we can still leave this issue be until it can be fixed. Say, if a massive refactoring is planned anyway, we can fix this as well. Then perhaps we should add a comment explaining situation and change the comment in a way that it does not pretend that this is a real proof meanwhile.

**Comment by bkchr** [comment]
> Correct me if I am wrong, but the proof there is a complete bogus. It, apparently, relies on the `GrandpaApi_grandpa_authorities` to return a non-empty list. The client code cannot rely on this particular behavior, because, well as was proven by the OP's counter-example, nothing stops runtime to return anything. Hence, the proof is wrong and the code needs to be changed.

Without having checked the code, I assume that you could also return an error. The point here being, that you *NEED* to have authorities in genesis, if not it is wrong/a failure or call it whatever you want. We have certain places where we have such panics, where if something is not that way we can only panic because it is just wrong.

**Comment by pepyakin** [comment]
Yeah, I didn't mean to say that this is a correct condition, just that the panic seems to be inappropriate and a graceful error should be used instead.

**Comment by andresilva** [comment]
We could turn this into an error, but we'll end up doing anyway is panic once we see that error. I'd prefer that instead we use some kind of `NonEmptyVec` type for such cases.
